govcd/user.go

```diff
@@ -0,0 +1,463 @@
+}
+
+const (
+	// Common role names and provider types are kept here to reduce hard-coded text and prevent mistakes
+	// Roles that are added to the organization need to be entered as free text
+
+	OrgUserRoleOrganizationAdministrator = "Organization Administrator"
+	OrgUserRoleCatalogAuthor             = "Catalog Author"
+	OrgUserRoleVappAuthor                = "vApp Author"
+	OrgUserRoleVappUser                  = "vApp User"
+	OrgUserRoleConsoleAccessOnly         = "Console Access Only"
+	OrgUserRoleDeferToIdentityProvider   = "Defer to Identity Provider"
+
+	// Allowed values for provider types
+	OrgUserProviderIntegrated = "INTEGRATED" // The user is created locally or imported from LDAP
+	OrgUserProviderSAML       = "SAML"       // The user is imported from a SAML identity provider.
+	OrgUserProviderOAUTH      = "OAUTH"      // The user is imported from an OAUTH identity provider
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		adminOrg: org,
+	}
+}
+
+// GetUserByNameOrId retrieves an user within an admin organization
+// by either name or ID
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+func (adminOrg *AdminOrg) GetUserByNameOrId(identifier string, willRefresh bool) (*OrgUser, error) {
+	if willRefresh {
+		err := adminOrg.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, orgUser := range adminOrg.AdminOrg.Users.User {
+		if (orgUser.Name == identifier || orgUser.ID == identifier) && orgUser.Type == types.MimeAdminUser {
+
+			user := NewUser(adminOrg.client, adminOrg)
+
+			_, err := adminOrg.client.ExecuteRequest(orgUser.HREF, http.MethodGet,
+				types.MimeAdminUser, "error getting user: %s", nil, user.User)
+
+			return user, err
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetRole finds a role within the organization
+func (adminOrg *AdminOrg) GetRole(roleName string) (*types.Reference, error) {
+
+	// There is no need to refresh the AdminOrg, until we implement CRUD for roles
+	for _, role := range adminOrg.AdminOrg.RoleReferences.RoleReference {
+		if role.Name == roleName {
+			return role, nil
+		}
+	}
+
+	return &types.Reference{}, ErrorEntityNotFound
+}
+
+// CreateUser creates an OrgUser from a full configuration structure
+// The timeOut variable is the maximum time we wait for the user to be ready
+// (This operation does not return a task)
+// Note that the timeout is not absolute, meaning that we only wait the
+// full amount of the timeout if the user has not been created until then.
+// This function returns as soon as the user has been created, which could be as
+// little as 50ms or as much as TimeOut - 50 ms
```

Comment:
"as little as 50ms or as much as TimeOut - 50 ms" - please rephrase, not clear why "50" is used twice.

Reply:
The first 50ms is the minimum. The second one is used as a value to subtract from the maximum.

If timeout is 10 seconds, a successful operation can take from 50 ms (the minimum) to 9 seconds and 950 ms (= 10 seconds - 50 ms). It would then fail at 10 seconds.

Reply:
Ah! Heh, had no clue that this hyphen represents a minus ;) Please change "-" to "minus" and "50 ms" to "50ms" OR just drop the last "- 50 ms" part.

Reply:
Removed because simplified approach